### PR TITLE
Add a preferences window

### LINKED ISSUE #2166: Create Centralized Preferences Window
**Is your feature request related to a problem? Please describe.**
Many options and defaults are changed and set in different places throughout SasView.

**Describe the solution you'd like**
There should be a single, cohesive window that handles all preferences and options. This panel should tie into the config file(s) to set defaults as needed.

**Additional context**
This is related to #258, #268, and #1726. There are likely more issues that I wasn't able to find in a 2-minute search.

### PR DESCRIPTION
**Pull request by krzywon**
This adds a preferences panel that can be opened from the File menu. There is a helper class for adding property widgets, a help file, and some basic unit tests.

I'm opening this as a draft because @lucas-wilkins is actively working on the config files for #2164 and we plan to combine our efforts once both are ready.

Fixes #2166

### CODE REVIEW

**Comment by lucas-wilkins** [comment]
See https://github.com/SasView/sasview/issues/2176 for why your tests are failing, merge the associated branch for fix, or merge main when https://github.com/SasView/sasview/pull/2178 is merged in.

**Comment by lucas-wilkins** [comment]
Hey, I'm pretty close to having the config stuff updated. There's a lot of changes, but it should be pretty simple to use. Just ironing out bugs that are only detected when building the installers.

File: src/sas/qtgui/Utilities/PreferencesPanel.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        super(PreferencesPanel, self).__init__(parent)
+        self.setupUi(self)
+        self.parent = parent
+        self.setWindowTitle("Preferences")
+        self.warning = None
+        # A list of callables used to restore the default values for each item in StackedWidget
+        self.restoreDefaultMethods = []
+        # Set defaults values for the list and stacked widgets
+        self.stackedWidget.setCurrentIndex(0)
+        self.listWidget.setCurrentRow(0)
```
Comment (rozyczko) [inline src/sas/qtgui/Utilities/PreferencesPanel.py:61]: why is this necessary in the `__init__` method? Aren't current indices/rows set to 0 anyway on instantiation?

File: src/sas/qtgui/Utilities/PreferencesPanel.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+    def onClick(self, btn: QPushButton):
+        """Handle button click events in one area"""
+        # Reset to the default preferences
+        if btn.text() == 'Restore Defaults':
+            self.restoreDefaultPreferences()
+        elif btn.text() == 'OK':
+            self.close()
+        elif btn.text() == 'Help':
+            self.help()
```
Comment (rozyczko) [inline src/sas/qtgui/Utilities/PreferencesPanel.py:79]: it's probably better to create separate slots for all the buttons rather than compare with the `text` property. Those texts can change and updating the handlers is not an obvious follow up.

File: src/sas/qtgui/Utilities/PreferencesPanel.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        for method in self.restoreDefaultMethods:
+            if callable(method):
+                method()
+            else:
+                logger.warning(f'While restoring defaults, {str(method)} of type {type(method)}'
+                               + ' was given. A callable object was expected.')
+
+    def close(self):
+        """Save the configuration values when the preferences window is closed"""
+        if hasattr(self.parent, 'guiManager'):
```
Comment (rozyczko) [inline src/sas/qtgui/Utilities/PreferencesPanel.py:92]: Why is this check necessary? Isn't `guiManager` always present on the mainWindow?

File: src/sas/qtgui/Utilities/PreferencesPanel.py (hunk 1 above)
Comment (krzywon) [reply src/sas/qtgui/Utilities/PreferencesPanel.py:61]: It's not necessary, but a good practice to ensure the two widgets are at the same point.

File: src/sas/qtgui/Utilities/PreferencesPanel.py (hunk 2 above)
Comment (krzywon) [reply src/sas/qtgui/Utilities/PreferencesPanel.py:79]: Agreed. This was a placeholder.

File: src/sas/qtgui/Utilities/PreferencesPanel.py (hunk 3 above)
Comment (krzywon) [reply src/sas/qtgui/Utilities/PreferencesPanel.py:92]: For unit testing, I create a dummy parent with a minimal implementation of the main window that doesn't have `guiManager`. When #2168 is finished, how the custom config is written will change so the check will not be required.

**Comment by wpotrzebowski** [comment]
@krzywon is this PR still in DRAFT state?

**Comment by krzywon** [comment]
> @krzywon is this PR still in DRAFT state?

This is waiting on #2168 to be finalized. Once it is, I will need to make a few changes here to account for the configuration changes.

**Comment by krzywon** [comment]
I have updated the window to account for changes made to the configuration system in PR #2168. There are a few outstanding modifications needed, as noted by @rozyczko, otherwise, this is ready for review.

**Question:** There are at least three ways forward from here - which one is preferred? I plan to work on 2 and 3 (and beyond) during the code camp if that makes a difference:

1. I can keep this PR as bare as possible with just the blank window. Other settings can then be moved into the window by submitting separate PRs.
2. I can add to this PR by creating a basic widget that houses the existing configuration values. (default file extensions, open folders, toolbar visibility, etc.)
3. Along with (2), I can also move certain preferences into this window, including the fitting options, and the fitting optimizer selection tool.

**Comment by lucas-wilkins** [comment]
Regarding option 2:

> I can add to this PR by creating a basic widget that houses the existing configuration values. (default file extensions, open folders, toolbar visibility, etc.)

I'm not sure how many of these need to be in a preferences panel. I fact, I'm not sure some of them belong in the config, I just wasn't sure about removing them from it at the time.

**Review by wpotrzebowski — APPROVED** [review]
I think it is good enough to give basis for Preferences windows. I've vote for merging it and move on with remaining functionality later

**Comment by rozyczko** [comment]
One potential issue is with multiple config files design. Should we not expect tab-per-config file view rather than have all config items thrown into a single window/view?
On the other hand, this GUI should offer a view on the json file generated by SasView rather than on the python files with definitions/defaults, so it's probably moot.

**Comment by krzywon** [comment]
I merged main, updated my unit tests, and added tests for the preferences widget generator. Assuming CI succeeds, this should now be ready to merge
